### PR TITLE
JAMES-3202 reindex outdated indexes

### PR DESCRIPTION
**Pull request by Arsnael**

### CODE REVIEW

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java [hunk 1]
```diff
@@ -1,150 +1,159 @@
...
-        return index.update(session, mailbox, updatedFlagsList);
-    }
-    
-
-    @Override
-    public Flux<MessageId> search(MailboxSession session, Collection<MailboxId> mailboxIds, SearchQuery searchQuery, long limit) throws MailboxException {
-        throw new UnsupportedSearchException();
-    }
-}
+/****************************************************************
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/LazyMessageSearchIndex.java:1]: There likely is a linebreak change here, be sure to commit it separately to ease review (sorry)

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearchIndex.java [hunk 2]
```diff
@@ -55,6 +58,11 @@
 
     EnumSet<MailboxManager.SearchCapabilities> getSupportedCapabilities(EnumSet<MailboxManager.MessageCapabilities> messageCapabilities);
 
+    /**
+     * Retrieves flags of an indexed message
+     */
+    Mono<Flags> retrieveIndexedFlags(MailboxSession session, MailboxId mailboxId, MessageUid uid) throws MailboxException;
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearchIndex.java:64]: Retrieving indexed data likely don't make search for the scanning implementation, it should be defined in `ListeningMessageSearchIndex`

Also I would make session parameter the last to keep consistent with the implicit mailbox-api convention.

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java [hunk 3]
```diff
@@ -1,178 +1,190 @@
...
-    }
-
-    private Flux<MessageId> getAsMessageIds(Flux<? extends SearchResult> temp, long limit) {
-        return temp.map(searchResult -> searchResult.getMessageId().get())
-            .filter(SearchUtil.distinct())
-            .take(Long.valueOf(limit).intValue());
-    }
-
-}
+/****************************************************************
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:1]: Idem, line delimiters changes should be in a separate commit.

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/RunningOptionsDTO.java [hunk 4]
```diff
@@ -28,22 +28,33 @@
...
 
+    public Optional<RunningOptions.Mode> getMode() {
+        return mode;
+    }
+
     public RunningOptions toDomainObject() {
-        return new RunningOptions(messagesPerSecond.orElse(RunningOptions.DEFAULT.getMessagesPerSecond()));
+        return new RunningOptions(
+            messagesPerSecond.orElse(RunningOptions.DEFAULT.getMessagesPerSecond()),
+            mode.orElse(RunningOptions.DEFAULT.getMode()));
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/RunningOptionsDTO.java:57]: I think we need a separated DEFAULT_MODE constant that RunningOptions.DEFAULT would then depend on.

File: mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ErrorRecoveryIndexationTaskSerializationTest.java [hunk 5]
```diff
@@ -43,10 +43,10 @@
...
+        " \"previousFailures\" : [{\"mailboxId\":\"1\",\"uids\":[10]},{\"mailboxId\":\"2\",\"uids\":[20]}], \"runningOptions\":{\"messagesPerSecond\":50, \"mode\":\"rebuild\"}}";
     private final String legacySerializedErrorRecoveryReindexingTask = "{\"type\": \"error-recovery-indexation\"," +
         " \"previousFailures\" : [{\"mailboxId\":\"1\",\"uids\":[10]},{\"mailboxId\":\"2\",\"uids\":[20]}]}";
-    private final String serializedAdditionalInformation = "{\"type\": \"error-recovery-indexation\", \"successfullyReprocessedMailCount\":42,\"failedReprocessedMailCount\":2,\"failures\":[{\"mailboxId\":\"1\",\"uids\":[10]},{\"mailboxId\":\"2\",\"uids\":[20]}], \"timestamp\":\"2018-11-13T12:00:55Z\", \"runningOptions\":{\"messagesPerSecond\":50}}";
+    private final String serializedAdditionalInformation = "{\"type\": \"error-recovery-indexation\", \"successfullyReprocessedMailCount\":42,\"failedReprocessedMailCount\":2,\"failures\":[{\"mailboxId\":\"1\",\"uids\":[10]},{\"mailboxId\":\"2\",\"uids\":[20]}], \"timestamp\":\"2018-11-13T12:00:55Z\", \"runningOptions\":{\"messagesPerSecond\":50, \"mode\":\"rebuild\"}}";
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/test/java/org/apache/mailbox/tools/indexer/ErrorRecoveryIndexationTaskSerializationTest.java:49]: No tests for the other mode?

File: server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRunningOptionsParser.java [hunk 6]
```diff
@@ -28,9 +28,9 @@
 public class ReindexingRunningOptionsParser {
 
     public static RunningOptions parse(Request request) {
-        return intQueryParameter(request, "messagesPerSecond")
-            .map(RunningOptions::new)
-            .orElse(RunningOptions.DEFAULT);
+        return new RunningOptions(
+            intQueryParameter(request, "messagesPerSecond").orElse(RunningOptions.DEFAULT.getMessagesPerSecond()),
+            modeQueryParameter(request, "mode").orElse(RunningOptions.DEFAULT.getMode()));
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/main/java/org/apache/james/webadmin/routes/ReindexingRunningOptionsParser.java:33]: We should consider adding a builder to these running options, so that we better handle default values.

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java [hunk 7]
```diff
@@ -29,17 +29,29 @@
...
     class RunningOptions {
-        public static final RunningOptions DEFAULT = new RunningOptions(50);
+        public enum Mode {
+            rebuild,
+            correct
+        }
+
+        public static final RunningOptions DEFAULT = new RunningOptions(50, Mode.rebuild);
 
         private final int messagesPerSecond;
```
Comment (blackheaven) [inline mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:40]: extra line

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java [hunk 8]
```diff
@@ -1,178 +1,190 @@
...
+    private static UidCriterion findConjugatedUidCriterion(List<Criterion> crits) {
+        for (Criterion crit : crits) {
+            if (crit instanceof UidCriterion) {
+                return (UidCriterion) crit;
+            } else if (crit instanceof ConjunctionCriterion) {
+                return findConjugatedUidCriterion(((ConjunctionCriterion) crit)
+                        .getCriteria());
+            }
+        }
+        return null;
```
Comment (blackheaven) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:112]: `Optional.empty()`

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java [hunk 9]
```diff
@@ -1,178 +1,190 @@
...
+        if (uidCrit != null) {
+            // if there is a conjugated uid range criterion in the query tree we can optimize by
+            // only fetching this uid range
+            UidRange[] ranges = uidCrit.getOperator().getRange();
+            for (UidRange r : ranges) {
+                Iterator<MailboxMessage> it = mapper.findInMailbox(mailbox, MessageRange.range(r.getLowValue(), r.getHighValue()), FetchType.Metadata, UNLIMITED);
+                while (it.hasNext()) {
+                    hitSet.add(it.next());
+                }
+            }
```
Comment (blackheaven) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:139]: use streams

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java [hunk 10]
```diff
@@ -1,178 +1,190 @@
...
+        } else {
+            // we have to fetch all messages
+            Iterator<MailboxMessage> messages = mapper.findInMailbox(mailbox, MessageRange.all(), FetchType.Full, UNLIMITED);
+            while (messages.hasNext()) {
+                MailboxMessage m = messages.next();
+                hitSet.add(m);
+            }
```
Comment (blackheaven) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:146]: idem

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java (hunk 8 above)
Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:112]: Legacy code, nothing to do with my changes

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java (hunk 9 above)
Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:139]: Legacy code, nothing to do with my changes

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java (hunk 10 above)
Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/search/SimpleMessageSearchIndex.java:146]: Legacy code, nothing to do with my changes

**Comment by LanKhuat** [comment]
Read it

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java [hunk 11]
```diff
@@ -264,4 +265,29 @@ void deleteMessagesShouldNotThrowWhenEmptyList() {
...
+
+        GetResponse getResponse = testee.get(documentId, useDocumentId(documentId)).block();
+
+        assertThat(getResponse.getSourceAsString()).isEqualTo(content);
+    }
+
+    @Test
+    void getShouldThrowWhenIdIsNull() {
+        assertThatThrownBy(() -> testee.get(null, ROUTING).block())
+            .isInstanceOf(NullPointerException.class);
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java:285]: Use `blockOptional` along with `.isEmtpy()` check ;-)

File: backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java [hunk 12]
```diff
@@ -264,4 +265,29 @@ void deleteMessagesShouldNotThrowWhenEmptyList() {
...
+    }
+
+    @Test
+    void getShouldThrowWhenRoutingKeyIsNull() {
+        assertThatThrownBy(() -> testee.get(DOCUMENT_ID, null).block())
+            .isInstanceOf(NullPointerException.class);
```
Comment (chibenwa) [inline backends-common/elasticsearch/src/test/java/org/apache/james/backends/es/ElasticSearchIndexerTest.java:291]: Idem

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java [hunk 13]
```diff
@@ -29,17 +31,60 @@
...
+            }
+        }
+
+        public static Builder builder() {
+            return new Builder();
+        }
+
+        public enum Mode {
+            rebuild,
+            correct
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:67]: We tend to use capital letters for enum. If you want to accept variations, you might want to write a parse static method.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 14]
```diff
@@ -213,4 +217,40 @@ private void logIfNoMessageId(SearchResult searchResult) {
...
+    }
+
+    private Flags extractFlags(Map<String, Object> source) {
+        Flags flags = new Flags();
+
+        if ((Boolean) source.get(JsonMessageConstants.IS_ANSWERED)) {
+            flags.add(Flags.Flag.ANSWERED);
+        }
+        if ((Boolean) source.get(JsonMessageConstants.IS_DELETED)) {
+            flags.add(Flags.Flag.DELETED);
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:240]: This mutability is not nice. Use `FlagsBuilder` instead.

I propose to modify it to also allow this syntax:

```
return FlagsBuilder.builder()
    .isAnswered(boolean)
    .isDeleted(boolean)
    .isDraft(boolean)
    ...
    .build();
```

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java [hunk 15]
```diff
@@ -213,4 +217,40 @@ private void logIfNoMessageId(SearchResult searchResult) {
...
+                String.format("Index for message %s in mailbox %s not found", uid.toString(), mailbox.getMailboxId().serialize()))));
+    }
+
+    private Flags extractFlags(Map<String, Object> source) {
+        Flags flags = new Flags();
+
+        if ((Boolean) source.get(JsonMessageConstants.IS_ANSWERED)) {
+            flags.add(Flags.Flag.ANSWERED);
+        }
+        if ((Boolean) source.get(JsonMessageConstants.IS_DELETED)) {
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndex.java:239]: Extract this condition in a generic method. Don't repeat yourself.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java [hunk 16]
```diff
@@ -442,4 +445,65 @@ void deleteAllShouldNotThrowWhenEmptyIndex() {
...
+        flags.add(Flags.Flag.RECENT);
+        flags.add(Flags.Flag.DRAFT);
+
+        SimpleMailboxMessage message = MESSAGE_BUILDER.messageId(MESSAGE_ID_4)
+            .uid(MESSAGE_UID_4)
+            .flags(flags)
+            .build();
+
+        testee.add(session, mailbox, message).block();
+        elasticSearch.awaitForElasticSearch();
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java:460]: Simple question: can you confirm if these awaits are needed?

File: mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java [hunk 17]
```diff
@@ -668,4 +668,14 @@ void notOperatorShouldReverseMatching() throws Exception {
         Stream<MessageUid> result = index.search(session, mailbox, query);
         assertThat(result).containsExactly(uid3, uid4);
     }
+
+    @Test
+    void retrieveIndexedFlagsShouldRetrieveFlags() {
```
Comment (chibenwa) [inline mailbox/lucene/src/test/java/org/apache/james/mailbox/lucene/search/LuceneMailboxMessageSearchIndexTest.java:673]: IMO we should have the same "contract" for `retrieveIndexedFlags` between ElasticSearch & Lucene.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java (hunk 16 above)
Comment (Arsnael) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java:460]: Seems not really but maybe safer to flush the indices, before trying to do a get operation on it?

**Comment by Arsnael** [comment]
Had to rebase on master to fix a conflict

**Comment by Arsnael** [comment]
There is integration test missing on `UserMailboxesRoutesTest` but taking out those #@$!* mocks is a pain... Will do next time. Might have unit tests missing around too, don't hesitate to give feedback.

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java (hunk 16 above)
Comment (chibenwa) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java:462]: This depends, documents don't need to be indexed to be get, I wonder if the flush is needed.

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 18]
```diff
@@ -255,7 +256,38 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
...
+            });
+    }
+
+    private boolean isIndexUpToDate(MailboxMessage message, Flags flags) {
+        return message.isAnswered() == flags.contains(Flags.Flag.ANSWERED)
+            && message.isRecent() == flags.contains(Flags.Flag.RECENT)
+            && message.isDeleted() == flags.contains(Flags.Flag.DELETED)
+            && message.isFlagged() == flags.contains(Flags.Flag.FLAGGED)
+            && message.isDraft() == flags.contains(Flags.Flag.DRAFT)
+            && message.isSeen() == flags.contains(Flags.Flag.SEEN);
```
Comment (chibenwa) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:287]: How about user flags?

How about using `message.createFlags()`

File: mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java (hunk 16 above)
Comment (Arsnael) [reply mailbox/elasticsearch/src/test/java/org/apache/james/mailbox/elasticsearch/events/ElasticSearchListeningMessageSearchIndexTest.java:462]: Seems working without so let's try it out

File: server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java [hunk 19]
```diff
@@ -1284,6 +1346,75 @@ void userReprocessingShouldReturnTaskDetailsWhenFailingAtTheMailboxLevel() throw
...
+                .then()
+                    .body("status", Matchers.is("completed"))
+                    .body("taskId", Matchers.is(notNullValue()))
+                    .body("type", Matchers.is(UserReindexingTask.USER_RE_INDEXING.asString()))
+                    .body("additionalInformation.successfullyReprocessedMailCount", Matchers.is(2))
+                    .body("additionalInformation.failedReprocessedMailCount", Matchers.is(0))
+                    .body("startedDate", Matchers.is(notNullValue()))
+                    .body("submitDate", Matchers.is(notNullValue()))
+                    .body("completedDate", Matchers.is(notNullValue()));
+            }
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: Do additionalInformation.successfullyReprocessedMailCount includes mails that did not had inconsistencies (I expect it should)

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: I'm missing the following test:

 - Verify in ElasticSearch that an inconsistency is fixed after running the task
 - Verify in ElasticSearch that nothing changes when there is no inconsistency after running the task.
 - Verify that "orphan" messages in ElasticSearch are well deleted. (or at least document the limitation)

Comment (Arsnael) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: > Do additionalInformation.successfullyReprocessedMailCount includes mails that did not had inconsistencies (I expect it should)

Mails that did not have inconsistencies are not being reprocessed, so I would on the contrary expect that it should not be included in the count of mails reprocessed?

Comment (Arsnael) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: > Verify that "orphan" messages in ElasticSearch are well deleted. (or at least document the limitation) 

The limitation has been documented: https://github.com/linagora/james-project/pull/3429/commits/823ad41cdc0bb97d4c6028bacb558152fdf4bc7b

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: > Mails that did not have inconsistencies are not being reprocessed, so I would on the contrary expect that it should not be included in the count of mails reprocessed?

I would rather expect to differenciate `processedMails` from `fixedInconsistencies`, no?

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: >  The limitation has been documented: 823ad41

A @Ignored test is the best documentation you could ever write for this ;-)

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailbox/src/test/java/org/apache/james/webadmin/routes/UserMailboxesRoutesTest.java:1417]: Should we apply the throttling on all emails? Or only on the ones having inconsistencies?

In other tasks we applied that for all emails.

**Comment by chibenwa** [comment]
```

[f38707de854be45f201169a1132e51ab2bd9b7ce] [ERROR]   ElasticSearchIndexerTest.getShouldThrowWhenIdIsNull:284 » NullPointer
[f38707de854be45f201169a1132e51ab2bd9b7ce] [ERROR]   ElasticSearchIndexerTest.getShouldThrowWhenRoutingKeyIsNull:290 » NullPointer
```

Oups. Sounds related.

**Comment by Arsnael** [comment]
Work has been split in two. Other part in -> https://github.com/linagora/james-project/pull/3452

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java [hunk 20]
```diff
@@ -29,17 +31,65 @@
...
+            }
+        }
+
+        public static Builder builder() {
+            return new Builder();
+        }
+
+        public enum Mode {
+            REBUILD,
+            CORRECT
```
Comment (remk) [inline mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:67]: I am not a fan of 'CORRECT' Mode as we can too easily be confused when reading the code as we can read it as an adjective on Mode.
I would propose `FIX`.

Or the more descriptive : 
`REBUILD_ALL`
`FIX_OUTDATED` / `CORRECT_OUTDATED`

**Comment by Arsnael** [comment]
```
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR] Errors: 
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   MailboxesRoutesTest$FullReprocessing$TaskDetails.fullReprocessingWithCorrectModeShouldRemoveOrphanMessagesInES:486 » NullPointer
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   MailboxesRoutesTest$MailboxReprocessing$TaskDetails.mailboxReprocessingWithCorrectModeShouldRemoveOrphanMessagesInES:864 » NullPointer
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   UserMailboxesRoutesTest$UserReprocessing$TaskDetails.userReprocessingWithCorrectModeShouldRemoveOrphanMessagesInES:1507 » NullPointer
```
Well expected as it's the current limitation... did put a `@ignore` tag, any idea why those tests are still running?

**Comment by chibenwa** [comment]
If it is junit 5 you need a @Disabled

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java (hunk 20 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:67]: I agree with FIX_OUTDATED wording

**Comment by chibenwa** [comment]
```
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR] Failures: 
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   MailboxesRoutesTest$FixingReIndexing$TaskDetails.userReprocessingShouldReturnTaskDetailsWhenFailingAtTheMailboxLevel:1304 1 expectation failed.
[6b881c74bda8296a5445e27eecf81bc02a4ff131] JSON path status doesn't match.
[6b881c74bda8296a5445e27eecf81bc02a4ff131] Expected: is "failed"
[6b881c74bda8296a5445e27eecf81bc02a4ff131]   Actual: null
[6b881c74bda8296a5445e27eecf81bc02a4ff131] 
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR] Errors: 
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   MailboxesRoutesTest$FullReprocessing$TaskDetails.fullReprocessingWithCorrectModeShouldRemoveOrphanMessagesInES:486 » NullPointer
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   MailboxesRoutesTest$MailboxReprocessing$TaskDetails.mailboxReprocessingWithCorrectModeShouldRemoveOrphanMessagesInES:864 » NullPointer
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR]   UserMailboxesRoutesTest$UserReprocessing$TaskDetails.userReprocessingWithCorrectModeShouldRemoveOrphanMessagesInES:1507 » NullPointer
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [INFO] 
[6b881c74bda8296a5445e27eecf81bc02a4ff131] [ERROR] Tests run: 394, Failures: 1, Errors: 3, Skipped: 2
```

**Comment by Arsnael** [comment]
Fixed a rebase issue with master and also added a few other fixups

File: mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java [hunk 21]
```diff
@@ -218,7 +218,7 @@ public ReIndexerPerformer(MailboxManager mailboxManager,
                 .flatMap(this::createReindexingEntryFromFailure),
             Flux.fromIterable(previousReIndexingFailures.mailboxFailures())
                 .flatMap(mailboxId -> mapper.findMailboxByIdReactive(mailboxId)
```
Comment (remk) [inline mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:220]: don't we need MAILBOX_CONCURRENCY on this flatmap ?

Comment (chibenwa) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:220]: That should come with a rebase on master hopefully

Comment (Arsnael) [reply mailbox/tools/indexer/src/main/java/org/apache/mailbox/tools/indexer/ReIndexerPerformer.java:220]: Already the case, line 225

**Comment by Arsnael** [comment]
Rebased on master to solve conflicts

**Comment by Arsnael** [comment]
```
[9f658281e3eacaefd993272b75cdac38fa695af6] [ERROR] Failures: 
[9f658281e3eacaefd993272b75cdac38fa695af6] [ERROR]   RabbitMQEventDeadLettersIntegrationTest.groupIdOfFailedEventShouldBeStoredInDeadLetter:320 1 expectation failed.
[9f658281e3eacaefd993272b75cdac38fa695af6] JSON path . doesn't match.
[9f658281e3eacaefd993272b75cdac38fa695af6] Expected: iterable over ["org.apache.james.webadmin.integration.rabbitmq.RabbitMQEventDeadLettersIntegrationTest$RetryEventsListenerGroup"] in any order
[9f658281e3eacaefd993272b75cdac38fa695af6]   Actual: []
[9f658281e3eacaefd993272b75cdac38fa695af6] 
[9f658281e3eacaefd993272b75cdac38fa695af6] [ERROR] Errors: 
[9f658281e3eacaefd993272b75cdac38fa695af6] [ERROR]   RabbitMQDeletedMessageVaultIntegrationTest>DeletedMessageVaultIntegrationTest.vaultEndpointShouldRestoreImapDeletedMailbox:255 » ConditionTimeout
```
test this please

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java [hunk 22]
```diff
@@ -29,17 +32,81 @@
...
+            }
+        }
+
+        public static Optional<Mode> parseMode(String optionalMode) {
+            return Optional.ofNullable(optionalMode)
+                .flatMap(Mode::fromString);
+        }
+
+        public static final Mode DEFAULT_MODE = Mode.REBUILD_ALL;
+        public static final RunningOptions DEFAULT = new RunningOptions(50, DEFAULT_MODE);
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:93]: Use the builder here?

**Comment by chibenwa** [comment]
```
[f27c1370df49c9c195316781058a50bc27aa1218] [ERROR] fullReindexingShouldUpdateDetailsUponSingleMessageFullReadError{CassandraCluster}  Time elapsed: 0.229 s  <<< ERROR!
[f27c1370df49c9c195316781058a50bc27aa1218] java.lang.NoClassDefFoundError: Could not initialize class org.apache.james.mailbox.indexer.ReIndexer$RunningOptions
[f27c1370df49c9c195316781058a50bc27aa1218] 	at org.apache.mailbox.tools.indexer.CassandraReIndexerImplTest$FailureTesting.fullReindexingShouldUpdateDetailsUponSingleMessageFullReadError(CassandraReIndexerImplTest.java:323)
```

Oups?

**Comment by Arsnael** [comment]
> ```
> [f27c1370df49c9c195316781058a50bc27aa1218] [ERROR] fullReindexingShouldUpdateDetailsUponSingleMessageFullReadError{CassandraCluster}  Time elapsed: 0.229 s  <<< ERROR!
> [f27c1370df49c9c195316781058a50bc27aa1218] java.lang.NoClassDefFoundError: Could not initialize class org.apache.james.mailbox.indexer.ReIndexer$RunningOptions
> [f27c1370df49c9c195316781058a50bc27aa1218] 	at org.apache.mailbox.tools.indexer.CassandraReIndexerImplTest$FailureTesting.fullReindexingShouldUpdateDetailsUponSingleMessageFullReadError(CassandraReIndexerImplTest.java:323)
> ```
> 
> Oups?

Sounds like issues with the rebase... Infernal

**Comment by Arsnael** [comment]
```
[314c5073c1f8dc5c9ff37124f59fb324984a9a97] [ERROR] Errors: 
[314c5073c1f8dc5c9ff37124f59fb324984a9a97] [ERROR]   ConsistencyTasksIntegrationTest.shouldRecomputeMailboxCounters:268 » ConditionTimeout
```
test this please

File: src/site/markdown/server/manage-webadmin.md [hunk 23]
```diff
@@ -675,10 +701,22 @@ An admin can specify the concurrency that should be used when running the task:
 
 This optional parameter must have a strictly positive integer as a value and be passed as query parameter.
 
+An admin can also specify the reindexing mode it wants to use when running the task:
+
+ - `mode` the reindexing mode used. There is 2 for the moment:
```
Comment (remk) [inline src/site/markdown/server/manage-webadmin.md:706]: s/There is 2 for the moment/There are 2 modes for the moment

File: src/site/markdown/server/manage-webadmin.md [hunk 24]
```diff
@@ -813,10 +852,22 @@ An admin can specify the concurrency that should be used when running the task:
...
```
Comment (remk) [inline src/site/markdown/server/manage-webadmin.md:857]: s/There is 2 for the moment/There are 2 modes for the moment

File: src/site/markdown/server/manage-webadmin.md [hunk 25]
```diff
@@ -1002,10 +1054,22 @@ An admin can specify the concurrency that should be used when running the task:
...
```
Comment (remk) [inline src/site/markdown/server/manage-webadmin.md:1059]: s/There is 2 for the moment/There are 2 modes for the moment

File: mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java [hunk 26]
```diff
@@ -89,8 +89,10 @@ String getValue() {
                 .flatMap(Mode::fromString);
         }
 
-        public static final Mode DEFAULT_MODE = Mode.REBUILD_ALL;
-        public static final RunningOptions DEFAULT = builder().messagesPerSeconds(Optional.of(50))
+        private static final Mode DEFAULT_MODE = Mode.REBUILD_ALL;
+        private static final int DEFAULT_MESSAGES_PER_SECONDS = 50;
+
+        public static final RunningOptions DEFAULT = builder().messagesPerSeconds(Optional.of(DEFAULT_MESSAGES_PER_SECONDS))
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:95]: Hm I think it's because we are parsing request parameters from the http request, that are optional themselves. Is it really a bother?

[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:95]: Can't we add an 'int' version of it @Arsnael ?

[the remark this answers is not in the document]
Comment (Arsnael) [reply mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:95]: Or maybe you want to add other build options  without Optional?

[the remark this answers is not in the document]
Comment (remk) [reply mailbox/api/src/main/java/org/apache/james/mailbox/indexer/ReIndexer.java:95]: It's not a big deal, but i though that usually we avoid Optional in the builder methods. And switch to default in the build method.

**Comment by Arsnael** [comment]
Will fix the doc typos during the merge, if it's ok?

**Comment by chibenwa** [comment]
Merged
